src/HttpHeader.cc

```diff
@@ -1424,19 +1421,40 @@ HttpHeaderEntry::parse(const char *field_start, const char *field_end)
 
     if (name_len > 65534) {
         /* String must be LESS THAN 64K and it adds a terminating NULL */
-        debugs(55, DBG_IMPORTANT, "WARNING: ignoring header name of " << name_len << " bytes");
+        debugs(55, DBG_IMPORTANT, "WARNING: rejecting due to header name of " << name_len << " bytes");
         return NULL;
     }
 
-    if (Config.onoff.relaxed_header_parser && xisspace(field_start[name_len - 1])) {
+    /*
+     * RFC 7230 section 3.2.4:
+     * "No whitespace is allowed between the header field-name and colon.
+     * ...
+     *  A server MUST reject any received request message that contains
+     *  whitespace between a header field-name and colon with a response code
+     *  of 400 (Bad Request).  A proxy MUST remove any such whitespace from a
+     *  response message before forwarding the message downstream."
+     */
+    if (xisspace(field_start[name_len - 1])) {
+
+        if (msgType == hoRequest)
+            return nullptr;
+
+        // for now, also let relaxed parser remove this BWS from any non-HTTP messages
+        const bool stripWhitespace = (msgType == hoReply) ||
+                                     Config.onoff.relaxed_header_parser;
+        if (!stripWhitespace)
+            return nullptr; // reject if we cannot strip
+
         debugs(55, Config.onoff.relaxed_header_parser <= 0 ? 1 : 2,
                "NOTICE: Whitespace after header name in '" << getStringPrefix(field_start, field_end-field_start) << "'");
 
         while (name_len > 0 && xisspace(field_start[name_len - 1]))
             --name_len;
 
-        if (!name_len)
+        if (!name_len) {
+            debugs(55, DBG_IMPORTANT, "WARNING: rejecting due to header with name of " << name_len << " bytes");
```

Comment:
An average admin reading this message in cache.log would not know _what_ Squid is rejecting and what rejecting that thing actually _means_. Both answers are probably caller/context-dependent so this message itself is probably misplaced. I do not insist on fixing this in the PR because this PR does not introduce this problem and because the fix is probably not trivial.

The suggestion below just makes the wording less awkward without addressing the problems mentioned above. If you use it here, please apply to other similar messages touched by this PR, of course.

```suggestion
            debugs(55, DBG_IMPORTANT, "WARNING: cannot parse message headers due to a " << name_len << "-byte long header name");
```

Reply:
'cannot parse' would be wrong for this message. The header is just being rejected as invalid according to the HTTP requirements. The parser itself could continue on to other headers if we want to make that happen - out of scope now.

I am inclined to make this and the below output both level 2 so that the message only appears in the context of admin debugging why an HTTP message is being rejected. Do you agree?

Reply:
> 'cannot parse' would be wrong for this message

You are being too literal -- "parsing" can (and, at the admin level, usually should) be viewed as all the stages of internalizing a message, including RFC requirements checks, but please use whatever phrasing you think would be better than the current broken one.

> I am inclined to make this and the below output both level 2

Yes, that would be a step forward, especially if the corresponding access log entry makes it clear that the message was rejected (whatever that means). I am a bit worried about ICAP messages that do not have an enabled-by-default access log -- it may be difficult for the admin to know that something went wrong specifically when talking to the ICAP service, but we can hope that the side effects would be visible enough in most cases (and enhance this reporting after specific complains).